Backports for 1.5.2

KristofferC opened the pull request:
Backported PRs:
- [x] #36810 - Define `OrderStyle` for `Union{}`
- [x] #37130 - Fix string interpolation into Markdown.Table and Markdown.List
- [x] #37178 - fix #37126, spurious soft scope warning on hidden name
- [x] #37181 - fix #37134, error with macro returning `x::T -> y`
- [x] #37204 - fix #37199, indexing for 1-d views with offset ranges
- [x] #37315 - reset ACTIVE_PROJECT when building docs
- [x] #37357 - Mark that ismutable requires julia 1.5
- [x] #37385 - Make sure to build clangsa before running clangsa self tests
- [x] #37296 - Allow packages to force depwarns
- [x] #37404 - fix #37401, displaying `Enum` in the repl
- [x] #37149 - fix ccall of strlen
- [x] #37114 - Improve optimizations on NULL pointer check
- [x] #37365 - inference: fix precision regression
- [x] #37308 - fix #37282, printing juxtaposition of 0 confusion with hex,bin,oct literals
- [x] #37386 - fix a possible segfault during static compilation
- [x] #37394 - Build `julia_version.h` before running `clangsa`
- [x] #37285 - Only emit incoming values for the phi when needed
- [x] #37319 - codegen: phic/upsilon nodes can be set to undefined
- [x] #37388 - Fix use of CreateAlignedStore
- [x] #37443 - Fix `getfield_tfunc` for constant `TypeName` 
- [x] #37444 - Specialize _memory_offset on number of arguments.
- [x] #37261 - Fix type-equality involving `Type{Union{}}` 
- [x] #37207 - Prune egal types when forming a `Union` 
- [x] #37472  - Fix memory leak in umfpack_numeric! 
- [x] #37534 - fix "invalid assignment location" macro hygiene regression
- [x] #37511 - Fix getfield of potentially undef inline immutable field
- [x] bump Pkg to 1.5.2 

Contains multiple commits, manual intervention needed:
- [ ] #37243 - code coverage fixes

Non-merged PRs with backport label:
- [ ] #36821 - Fix function call to rec_backtrace_ctx when using USE_SYSTEM_LIBUNWIND=1

bkamins committed:
Define `OrderStyle` for `Union{}` (#36810)

Removes ambiguity:
```
julia> Base.OrderStyle(Union{})
ERROR: MethodError: Base.OrderStyle(::Type{Union{}}) is ambiguous. Candidates:
```
This error is relevant as with current `unique!` definition that relies on `OrderStyle` one can have a problem in corner cases. E.g.:
```
julia> [i for i in ["1"] if i isa Int]
0-element Array{Union{},1}
```

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>
(cherry picked from commit 03e1a89ff1f62f05cf3937f5da44f3d46956138a)

fredrikekre committed:
Fix string interpolation into Markdown.Table and Markdown.List in md"..." strings, fixes #16194. (#37130)

(cherry picked from commit d446c276c01d2b78973c3ea20310ed305656ace0)

JeffBezanson committed:
fix #37126, spurious soft scope warning on hidden name (#37178)

(cherry picked from commit 9ae0a73869cd4876b0b32ed78916e17cea290aeb)

yuyichao committed:
remove the incorrect return value attribute

yuyichao commented:
Seems that the back point of https://github.com/JuliaLang/julia/pull/37149 included more than just the bug fix. Only e change to gcutils.jl and secretbuffer.jl should be backported since those fixed the wrong return type from strlen call...

KristofferC commented:
Thanks, I missed that comment, will change it.

yuyichao committed:
fix ccall of strlen

JeffBezanson committed:
fix #37134, error with macro returning `x::T -> y` (#37181)

(cherry picked from commit 646626e32c924c3c3d8e99b2ab3c08e47d34ce62)

martinholters committed:
Prune egal types when forming a `Union` (#37207)

Fixes #37180.

(cherry picked from commit f047d7ffc7e7848d5a6ce9cc563043d7574f3e6a)

mbauman committed:
fix #37199, indexing for 1-d views with offset ranges (#37204)

We had been assuming that IdentityUnitRange matched the indices of the parent (like Slices) but they define their own offset axes. Further, other `AbstractRanges` may be similarly offset. We do not need to consider other offset `AbstractArrays` as this code only applies to `IndexLinear` `SubArray`s.

(cherry picked from commit 701885be8e72670fe7a759401ff20eeb3d41070d)

martinholters committed:
Fix type-equality involving `Type{Union{}}` (#37261)

In particular, make `obviously_unequal(Type{Union{}}, Type{T} where
{Union{}<:T<:Union{}})` false. Fixes #37255

(cherry picked from commit 79a9ea08c7ebe915af1f669b498c61a1d88b42c0)

KristofferC committed:
reset ACTIVE_PROJECT when building docs (#37315)

(cherry picked from commit f780151aceffa368de33e458f738467bf22e06d9)

JeffBezanson committed:
fix #37282, printing juxtaposition of 0 confusion with hex,bin,oct literals (#37308)

(cherry picked from commit f896006621414725ac1355fe23a779c50995091d)

yuyichao committed:
Only emit incoming values for the phi when needed (#37285)

Fixes a crash when one of the incoming value for a union value phi node is of type `Union{}`.
Incoming value of this type can be generated by try-catch.

Fix #37265

(cherry picked from commit c3a63fc591a178837445c3243fd91dcef1f8ec0e)

oxinabox committed:
Mark that ismutable requires julia 1.5 (#37357)

(cherry picked from commit 3453a5c4753b7baa9aed2ba32636ed088ce0c65e)

vtjnash committed:
inference: fix precision regression (#37365)

While in 1d08d70fb07cfcfc9e89ae1a2521411b1e1724e5 we wanted to avoid
combining two types, we did not intend to prevent a type from being just
a pass-through value.

(cherry picked from commit a9743d897d602ec899d2ff136a1fc4ad18da07ba)

vtjnash committed:
codegen: phic/upsilon nodes can be set to undefined (#37319)

Unlike other nodes, Upsilon nodes can be potentially undefined and need
special handling to ensure they aren't literally undef, as we might try
to union-copy or gc-root the data later.

Fix #37262

(cherry picked from commit 5cf571721369a83a072fbef8e3de8e20d7644aa8)

Keno committed:
Make sure to build clangsa before running clangsa self tests (#37385)

(cherry picked from commit f51d4cb04ecedb3db3a9fc99e60f93b124447625)

staticfloat committed:
Build `julia_version.h` before running `clangsa` (#37394)

(cherry picked from commit 315ad46670991a906fc0f381a2999999eb43a1dd)

yuyichao committed:
Fix use of CreateAlignedStore (#37388)

This is a compile error on 10+ and is probably triggering implicity conversion on LLVM 8 and 9.

(cherry picked from commit 5acd7bdfbb1b9deed9a9438f21eb8170fedeee57)

timholy committed:
Allow packages to force depwarns (#37296)

(cherry picked from commit 405cd6f67fa814c51c01d0f81e3320d2d5767386)

JeffBezanson committed:
fix #37401, displaying `Enum` in the repl (#37404)

(cherry picked from commit 483b63780a0ebd3141f30772a1d08b92d6fd7b78)

JeffBezanson committed:
fix a possible segfault during static compilation (#37386)

chriselrod committed:
Specialize _memory_offset on number of arguments. (#37444)


(cherry picked from commit 1e6d771d295b786e5ca4751d8e553c45b9e6c58f)

yuyichao committed:
Fix `getfield_tfunc` for constant `TypeName` (#37443)

The field should be either `Symbol` or `Int`. Ref #37423

Unlike #37423, in additional to worse type info in inference,
the missing type check here can actually cause type inference error
due to errors in user code.

(cherry picked from commit 1378bb6731e75469cbeacb31215bdc180c7ee913)

martinholters commented on test/compiler/inference.jl:
```diff
@@ -2617,3 +2625,25 @@ f36531(args...) = tuple((args...)...)
+for goodf in [getfield_const_typename_good1, getfield_const_typename_good2]
+    local goodf
+    local code = code_typed(goodf, Tuple{})[1].first.code
+    @test code[1] === Core.ReturnNode(QuoteNode(:Int32))
```
```suggestion
    @test code[1] === Core.Compiler.ReturnNode(QuoteNode(:Int32))
```

martinholters commented on test/compiler/inference.jl:
```diff
@@ -2617,3 +2625,25 @@ f36531(args...) = tuple((args...)...)
+    local badf
+    local code = code_typed(badf, Tuple{})[1].first.code
+    @test Meta.isexpr(code[1], :call)
+    @test code[end] === Core.ReturnNode()
```
```suggestion
    @test code[end] === Core.Compiler.ReturnNode()
```

martinholters reviewed (commented):
The tests from #37443 need a minor adjustment.

JeffBezanson commented:
There is a test failure here from 38c3e741af5e18d41f0420f5fe56a8a54b1ee248. It looks like some part of dcc0696971a100dbc63f90b8789632631dd1bfef is also needed to make that test work. That whole commit seems likely safe to backport though.

JeffBezanson commented:
This is the needed line:
```
--- a/base/compiler/inferencestate.jl
+++ b/base/compiler/inferencestate.jl
@@ -205,7 +205,7 @@ update_valid_age!(edge::InferenceState, sv::InferenceState) = update_valid_age!(
 function record_ssa_assign(ssa_id::Int, @nospecialize(new), frame::InferenceState)
     old = frame.src.ssavaluetypes[ssa_id]
     if old === NOT_FOUND || !(new ⊑ old)
-        frame.src.ssavaluetypes[ssa_id] = tmerge(old, new)
+        frame.src.ssavaluetypes[ssa_id] = old === NOT_FOUND ? new : tmerge(old, new)
         W = frame.ip
         s = frame.stmt_types
```

martinholters commented:
Ref. https://github.com/JuliaLang/julia/pull/37365#issuecomment-688721730 :wink:

KristofferC committed:
add a part of dcc0696971a1 needed to make tests pass

KristofferC commented:
@nanosoldier `runbenchmarks(ALL, vs = ":release-1.5")`

KristofferC commented:
@nanosoldier `runtests(ALL, vs = ":release-1.5")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/37210#issuecomment-689410445) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/40559b5cbc31f3c53c2e6af7cf9900fe70c42d06/be8475f_vs_c4acbf9/report.md). cc @ararslan

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/37210#issuecomment-689410535) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/801671c4f9782fadd4f9b6718570c91a1414fd0e/pkgeval/by_hash/21d877d_vs_c4acbf9/report.md). cc @maleadt

KristofferC commented:
https://github.com/JuliaCI/NanosoldierReports/blob/801671c4f9782fadd4f9b6718570c91a1414fd0e/pkgeval/by_hash/21d877d_vs_c4acbf9/logs/StatProfilerHTML/1.5.2-pre-be8475f41a.log

```
ERROR: LoadError: LoadError: syntax: invalid assignment location "#40#content..." around /home/pkgeval/.julia/packages/HAML/hMyIa/src/Templates.jl:49
Stacktrace:
 [1] top-level scope at none:1
 [2] eval at ./boot.jl:331 [inlined]
 [3] eval at ./Base.jl:39 [inlined]
 [4] _includehaml(::Module, ::Symbol, ::String, ::String) at /home/pkgeval/.julia/packages/HAML/hMyIa/src/Templates.jl:98
 [5] includehaml at /home/pkgeval/.julia/packages/HAML/hMyIa/src/Templates.jl:68 [inlined] (repeats 2 times)
 [6] #3 at /home/pkgeval/.julia/packages/HAML/hMyIa/src/Templates.jl:71 [inlined]
 [7] foreach at ./abstractarray.jl:2009 [inlined]
 [8] includehaml(::Module, ::Pair{Symbol,String}, ::Pair{Symbol,String}, ::Pair{Symbol,String}, ::Pair{Symbol,String}, ::Vararg{Pair{Symbol,String},N} where N) at /home/pkgeval/.julia/packages/HAML/hMyIa/src/Templates.jl:70
 [9] top-level scope at /home/pkgeval/.julia/packages/StatProfilerHTML/BuyY0/src/HTML.jl:12
 [10] include(::Function, ::Module, ::String) at ./Base.jl:380
 [11] include at ./Base.jl:368 [inlined]
 [12] include(::String) at /home/pkgeval/.julia/packages/StatProfilerHTML/BuyY0/src/StatProfilerHTML.jl:1
 [13] top-level scope at /home/pkgeval/.julia/packages/StatProfilerHTML/BuyY0/src/StatProfilerHTML.jl:6
 [14] include(::Function, ::Module, ::String) at ./Base.jl:380
```

looks a bit strange. @JeffBezanson?

cfauchereau committed:
Fix memory leak in umfpack_numeric! (#37472)

(cherry picked from commit b17cd101fc2e05a21b26a98fffcd20dd736e0653)

fredrikekre commented:
Backport of https://github.com/JuliaLang/julia/pull/37315 should be `HOME_PROJECT` on 1.5 (https://github.com/JuliaLang/julia/pull/37315#pullrequestreview-479690267)

JeffBezanson committed:
fix "invalid assignment location" macro hygiene regression (#37534)

(cherry picked from commit a83166aae7e98b2df886d6ac3f905a6aae8eaaeb)

yuyichao committed:
Fix getfield of potentially undef inline immutable field (#37511)

The undef ref error may be thrown at the wrong time or not at all and may cause crashes.

(cherry picked from commit 72854dc6e2874638cdd4f578bcfebc162ec347b5)

KristofferC committed:
bump Pkg to 1.5.2

KristofferC commented:
This should be good to go once CI finishes.
